src/ui/src/js/controllers/admin_user_view.js

```diff
@@ -0,0 +1,242 @@
+  $scope.alerts = [];
+
+  $scope.doDelete = function(userName) {
+    UserService.deleteUser(userName).then(
+        $state.go('base.user_admin'),
+        failureCallback,
+    );
+  };
+
+  $scope.closeAlert = function(index) {
+    $scope.alerts.splice(index, 1);
+  };
+
+  $scope.submitUserForm = function(form, model) {
+    clearScopeAlerts();
+    model = blankRequiredFields(model);
+    $scope.$broadcast('schemaFormValidate');
+
+    if (form.$valid) {
+      UserService.updateUser($scope.data.username, model).then(
+          addSuccessAlert,
+          addErrorAlert,
+      );
+    }
+  };
+
+  const blankRequiredFields = function(model) {
+    // ASF appears to have a bug when using required fields with arrays
+    // that makes it possible to get into a non-submittable state when the
+    // required fields are defined on the form. Instead, we require the fields
+    // on the model, and then just blank out the identifiers that shouldn't
+    // be present for the currently selected scope. The backend drops blank
+    // identifiers for us anyway.
+    model.role_assignments.forEach((assignment) => {
+      if (assignment.domain.scope === 'Global') {
+        assignment.domain.identifiers.name = '';
+      }
+      if (assignment.domain.scope !== 'System') {
+        assignment.domain.identifiers.namespace = '';
+        assignment.domain.identifiers.version = '';
+      }
+    });
+
+    return model;
+  };
+
+  const successCallback = function(response) {
+    $scope.response = response;
+    $scope.data = response.data;
+
+    generateUserSF();
+  };
+
+  const failureCallback = function(response) {
+    $scope.response = response;
+    $scope.data = [];
+  };
+
+  const addErrorAlert = function(response) {
+    $scope.alerts.push({
+      type: 'danger',
+      msg:
+        'Something went wrong on the backend: ' +
+        _.get(response, 'data.message', 'Please check the server logs'),
+    });
+  };
+
+  const addSuccessAlert = function(response) {
+    $scope.alerts.push({
+      type: 'success',
+      msg: 'Saved',
+    });
+  };
+
+  const clearScopeAlerts = function() {
+    while ($scope.alerts.length) {
+      $scope.alerts.pop();
+    }
+  };
+
+  const loadRoleNames = function() {
+    return RoleService.getRoles().then((response) => {
+      $scope.roleNames = _.map(response.data.roles, 'name');
+    });
+  };
+
+  const loadUser = function() {
+    $scope.response = undefined;
+    $scope.data = [];
+
+    return UserService.getUser($stateParams.username).then(
+        successCallback,
+        failureCallback,
+    );
+  };
+
+  const generateUserSF = function() {
+    $scope.userSchema = roleAssignmentSchema;
+    $scope.userForm = roleAssignmentForm;
+    $scope.userModel = serverModelToForm($scope.data);
+
+    // This gets set here because $scope.roleNames is not yet set at
+    // roleAssignmentSchema declaration time
+    $scope.userSchema.properties.role_assignments.items.properties.role_name.enum =
```

Comment:
The way the code is written, I believe it should be the case that the role names are guarenteed to be loaded by the time we get here (see the very last line of this file).  And then the form is redrawn immediately after this line.   Given that, what is supposed to happen is that the "Role" field is a dropdown populated by the actual list of roles that currently exist, per the `/api/v1/roles/` endpoint.  

For reasons that are totally unclear, this works like 80 - 90% of the time.  Every once in a while the role field will render as just a regular text input field rather than a dropdown.  I feel like this is a bug with the ASF library and not my code.  I tried a few variations on this, including defining the entire schema right here so that I didn't have to inject the enum property after the fact.  Nothing I tried fixed this inconsistent behavior.  If someone else has suggestions as to why they think that might be happening I'm open to trying to fix it.  Otherwise I'm chalking it up to an ASF bug and moving on.